services/image-rec-master/service/tasks.py

```diff
@@ -0,0 +1,383 @@
+                    tr = r.multi_exec()
+                    tr.lrem('unsubmitted-targets', 0, target_id)
+                    tr.lrem('removing-targets', 0, target_id)
+                    tr.sadd('removed-targets', target_id)
+                    tr.hset(target_key, 'removed', '1')
+                    logging.info(f'target {target_id} cancelled')
+
+                    await tr.execute()
+                    return
+        except aioredis.MultiExecError:
+            # The unsubmitted targets changed.
+            await asyncio.sleep(0.1)
+        except aioredis.RedisError as e:
+            logging.error(format_error('redis error', str(e)))
+            await asyncio.sleep(0.5)
+        else:
+            # Proceed to the next case.
+            break
+
+    r = app['redis']
+
+    # Wait for removal if the target is being submitted and,
+    # regardless, check if the target has been removed already.
+    while True:
+        try:
+            submitting = await get_int_list(r, 'submitting-targets')
+
+            # If the target is being submitted, wait and try it
+            # again.
+            if target_id in submitting:
+                await asyncio.sleep(0.5)
+                continue
+
+            # If the odlc was removed we don't need to do anything
+            # else to remove.
+            if await r.hget(target_key, 'removed') == b'1':
+                await r.lrem('removing-targets', 0, target_id)
+                return
+        except aioredis.RedisError as e:
+            logging.error(format_error('redis error', str(e)))
+            await asyncio.sleep(0.5)
+        else:
+            # Proceed to the next case.
+            break
+
+    # Otherwise, get the odlc id for removal.
+    while True:
+        try:
+            odlc = await r.hget(target_key, 'odlc')
+            odlc_id = Odlc.FromString(odlc).id
+        except aioredis.RedisError as e:
+            logging.error(format_error('redis error', str(e)))
+            await asyncio.sleep(0.1)
+        else:
+            break
+
+    # Remove the odlc from the interop server.
+    while True:
+        try:
+            removed = await _delete_odlc(app, odlc_id)
+        except aiohttp.ClientError as e:
+            # Handle the possible non-404 HTTP error above.
+            logging.error(format_error('http error', str(e)))
+            await asyncio.sleep(0.1)
+        else:
+            break
+
+    # Update Redis with the removal.
+    while True:
+        try:
+            tr = r.multi_exec()
+            tr.lrem('removing-targets', 0, target_id)
+
+            if removed:
+                tr.sadd('removed-targets', target_id)
+                tr.hset(target_key, 'removed', '1')
+                logging.info(f'target {target_id} removed')
+
+            await tr.execute()
+        except aioredis.RedisError as e:
+            logging.error(format_error('redis error', str(e)))
+            await asyncio.sleep(0.1)
+        else:
+            break
```

Comment:
Maybe we can clean up this `while True/except/else` pattern with a context manager?

Reply:
Added a story for it, I initially didn't do it since I wanted to keep the error handling specific to each case instead of the chance of unexpected exceptions from a block. But I don't think this is really a problem. Shouldn't be hard to generalize.